c/extras/py_ftrl.cc

```diff
@@ -0,0 +1,338 @@
+//------------------------------------------------------------------------------
+#include "extras/py_ftrl.h"
+#include "frame/py_frame.h"
+#include "python/float.h"
+#include "python/int.h"
+
+namespace py {
+
+PKArgs Ftrl::Type::args___init__(9, 0, 0, false, false,
+                                 {"a", "b", "l1", "l2", "d", "n_epochs",
+                                 "inter", "hash_type", "seed"},
+                                 "__init__", nullptr);
+
+void Ftrl::m__init__(PKArgs& args) {
+  double a = args[0].to_double();
+  double b = args[1].to_double();
+  double l1 = args[2].to_double();
+  double l2 = args[3].to_double();
+
+  uint64_t d = static_cast<uint64_t>(args[4].to_size_t());
+  size_t n_epochs = args[5].to_size_t();
+  bool inter = args[6].to_bool_strict();
+  unsigned int hash_type = static_cast<unsigned int>(args[7].to_size_t());
+  unsigned int seed = static_cast<unsigned int>(args[8].to_size_t());
+
+  fm = new FtrlModel(a, b, l1, l2, d, n_epochs, inter, hash_type, seed);
+}
+
+
+void Ftrl::Type::init_methods_and_getsets(Methods& mm, GetSetters& gs) {
+  gs.add<&Ftrl::get_model, &Ftrl::set_model>("model",
+    "Frame having two columns, i.e. `z` and `n`, and `d` rows,\n"
+    "where `d` is a number of bins set for modeling. Both column types\n"
+    "must be `FLOAT64`.\n"
+    "NB: as the model trains, this frame will be changed in-place.\n");
+
+  gs.add<&Ftrl::get_a, &Ftrl::set_a>("a", "`alpha` in per-coordinate learning rate formula.\n");
+  gs.add<&Ftrl::get_b, &Ftrl::set_b>("b", "`beta` in per-coordinate learning rate formula.\n");
+  gs.add<&Ftrl::get_l1, &Ftrl::set_l1>("l1", "L1 regularization parameter.\n");
+  gs.add<&Ftrl::get_l2, &Ftrl::set_l2>("l2", "L2 regularization parameter.\n");
+  gs.add<&Ftrl::get_d, &Ftrl::set_d>("d", "Number of bins to be used after the hashing trick.\n");
+  gs.add<&Ftrl::get_n_epochs, &Ftrl::set_n_epochs>("n_epochs", "Number of epochs to train for.\n");
+  gs.add<&Ftrl::get_inter, &Ftrl::set_inter>("inter", "If feature interactions to be used or not.\n");
+  gs.add<&Ftrl::get_hash_type, &Ftrl::set_hash_type>("hash_type",
+    "Hashing method to use for strings.\n"
+    "`0` - std::hash;\n"
+    "`1` - Murmur2;\n"
+    "`2` - Murmur3.\n");
+  gs.add<&Ftrl::get_seed, &Ftrl::set_seed>("seed", "Seed to be used for Murmur hash functions.\n");
+
+  mm.add<&Ftrl::fit, args_fit>();
+  mm.add<&Ftrl::predict, args_predict>();
+}
+
+
+PKArgs Ftrl::Type::args_fit(1, 0, 0, false, false, {"df_train"}, "fit",
+R"(fit(self, frame)
+--
+
+Train an FTRL model on a dataset.
+
+Parameters
+----------
+frame: Frame
+    Frame to be trained on, last column is treated as `target`.
+
+Returns
+----------
+    A new Frame of shape `(d, 2)` containing `z` and `n` model coefficients.
+)");
+
+
+void Ftrl::fit(const PKArgs& args) {
+  if (!args[0].is_frame()) {
+    throw TypeError() << "argument must be a frame, not "
+        << args[0].typeobj();
+  }
+  DataTable* dt_train = args[0].to_frame();
```

Comment:
this call will already perform the check `.is_frame()` and throw an exception with similar message as the above